File: IT/Multithreading/Esercizi/C++/esercizio24.cpp
/*
file esercizio24.cpp
compilare con: g++ -std=c++11 esercizio24.cpp -lpthread
*/

#include <random>
#include <thread>
#include <mutex>
#include <chrono>
#include <array>

using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::thread;
using std::recursive_mutex;
using std::chrono::seconds;
using std::array;

void philosopher(const char* name, array<std::recursive_mutex, 5>& forks, int rIndex, int lIndex) {
	thread_local default_random_engine gen(random_device{}());

	while (true) {
		if (forks[rIndex].try_lock()) {
			if (forks[lIndex].try_lock()) {
				printf("%s prende le forchette %d e %d.\n", name, rIndex, lIndex);
				printf("%s sta mangiando.\n", name);
				// ogni filosofo impiega tra 4 e 8 secondi per mangiare
				uniform_int_distribution<int> dist(4, 8);
				seconds duration(dist(gen));
				std::this_thread::sleep_for(duration);

				printf("%s posa le forchette %d e %d.\n", name, rIndex, lIndex);
				forks[lIndex].unlock();
			}

			forks[rIndex].unlock();
		}

		// ogni filosofo pensa per 3/5 secondi
		uniform_int_distribution<int> dist(3, 5);
		seconds duration(dist(gen));
		std::this_thread::sleep_for(duration);
		printf("%s sta pensando...\n", name);
	}
}

int main()
{
	const int N_PHILOSOPHERS = 5;
	const char* names[] = {"Platone", "Pitagora", "Kant", "Leibniz", "Marx"};

	array<recursive_mutex, 5> forks;
	thread philosophers[N_PHILOSOPHERS];

	for (int i = 0; i < N_PHILOSOPHERS; i++) {
		philosophers[i] = thread(philosopher, names[i], ref(forks), i, (i+1) % N_PHILOSOPHERS);
	}

	for (auto& thread : philosophers) {
		thread.join();
	}
}

File: IT/2/Multithreading/Esercizi/Java/Esercizio24.java
/* file Esercizio24.java */

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class Philosopher implements Runnable {
	private Lock [] forks;
	private int rIndex;
	private int lIndex;

	public Philosopher(Lock [] forks, int rIndex, int lIndex) {
		this.forks = forks;
		this.rIndex = rIndex;
		this.lIndex = lIndex;
	}

	@Override
	public void run() {
		Thread t = Thread.currentThread();
		while (true) {
			try {
				if (forks[rIndex].tryLock(10, TimeUnit.MILLISECONDS)) {
					if (forks[lIndex].tryLock(10, TimeUnit.MILLISECONDS)) {
						System.out.println(t.getName() + " prende le forchette " + rIndex + " e " + lIndex + ".");
						System.out.println(t.getName() + " sta mangiando.");

						// ogni filosofo impiega tra 4 e 8 secondi per mangiare
						int pause = ThreadLocalRandom.current().nextInt(4, 8+1);
						t.sleep(pause*1000);

						System.out.println(t.getName() + " posa le forchette " + rIndex + " e " + lIndex + ".");

						forks[lIndex].unlock();
					}

					forks[rIndex].unlock();
				}

				// ogni filosofo pensa per 3/5 minuti
				int pause = ThreadLocalRandom.current().nextInt(3, 5+1);
				t.sleep(pause*1000);
				System.out.println(t.getName() + " sta pensando...");
			}
			catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}
}

public class Esercizio24 {
	public static void main(String [] args) {
		final int N_PHILOSOPHERS = 5;
		String [] names = {"Platone", "Pitagora", "Kant", "Leibniz", "Marx"};
		Lock [] forks = new Lock[N_PHILOSOPHERS];

		for (int i = 0; i < N_PHILOSOPHERS; i++) {
			forks[i] = new ReentrantLock();
		}

		for (int i = 0; i < N_PHILOSOPHERS; i++) {
			Thread philosopher = new Thread(new Philosopher(forks, i, (i+1) % N_PHILOSOPHERS), names[i]);
			philosopher.start();
		}
	}
}

